pkg/email/lore/parse.go

```diff
@@ -33,24 +55,137 @@ func Threads(emails []*email.Email) []*Thread {
 	return ctx.threads
 }
 
+// PatchSeries is similar to Threads, but returns only the patch series submitted to the mailing lists.
+// TODO: add a test?
+func PatchSeries(emails []*email.Email) []*Series {
+	var ret []*Series
+	// Normally, all following series patches are sent in response to the first email sent.
+	// So there's no sense to look at deeper replies.
+	for _, thread := range listThreads(emails, 1) {
+		if thread.Type != dashapi.DiscussionPatch {
+			continue
+		}
+		patch, ok := parsePatchSubject(thread.Subject)
+		if !ok {
+			// It must never be happening.
+			panic("DiscussionPatch is set, but we fail to parse the thread subject")
+		}
+		series := &Series{
+			Subject:   patch.Title,
+			MessageID: thread.MessageID,
+			Total:     1,
+			Version:   1,
+		}
+		if patch.Total.IsSet() {
+			series.Total = patch.Total.Value()
+		}
+		if patch.Version.IsSet() {
+			series.Version = patch.Version.Value()
+		}
+		ret = append(ret, series)
+		if patch.Seq.IsSet() && patch.Seq.Value() > 1 {
+			series.Corrupted = "the first patch has seq>1"
+			continue
+		}
+		hasSeq := map[int]bool{}
+		for _, email := range thread.Messages {
+			patch, ok := parsePatchSubject(email.Subject)
+			if !ok {
+				continue
+			}
+			seq := 1
+			if patch.Seq.IsSet() {
+				seq = patch.Seq.Value()
+			}
+			if seq == 0 {
+				// The cover email is not of interest.
+				continue
+			}
+			if hasSeq[seq] {
+				// It's weird if that really happens, but let's skip for now.
+				continue
+			}
+			hasSeq[seq] = true
+			series.Patches = append(series.Patches, Patch{
+				Seq:   seq,
+				Email: email,
+			})
+		}
+		if len(hasSeq) != series.Total {
+			series.Corrupted = fmt.Sprintf("the subject mentions %d patches, %d are found",
+				series.Total, len(hasSeq))
+			continue
+		}
+		if len(series.Patches) == 0 {
+			series.Corrupted = fmt.Sprintf("0 patches")
+			continue
+		}
+		sort.Slice(series.Patches, func(i, j int) bool {
+			return series.Patches[i].Seq < series.Patches[j].Seq
+		})
+	}
+	return ret
+}
+
 // DiscussionType extracts the specific discussion type from an email.
 func DiscussionType(msg *email.Email) dashapi.DiscussionType {
 	discType := dashapi.DiscussionMention
 	if msg.OwnEmail {
 		discType = dashapi.DiscussionReport
 	}
 	// This is very crude, but should work for now.
-	if patchSubjectRe.MatchString(strings.ToLower(msg.Subject)) {
+	if _, ok := parsePatchSubject(msg.Subject); ok {
 		discType = dashapi.DiscussionPatch
 	} else if strings.Contains(msg.Subject, "Monthly") {
 		discType = dashapi.DiscussionReminder
 	}
 	return discType
 }
 
-var patchSubjectRe = regexp.MustCompile(`\[(?:(?:rfc|resend)\s+)*patch`)
+type PatchSubject struct {
+	Title   string
+	Tags    []string // Sometimes there's e.g. "net" or "next-next" in the subject.
```

Comment:
These tags are parsed, but lost later. What's the plan here?
Do we want to not test "RFC" patches? Then these should set Corrupted?

Reply:
As mentioned in the comment, the developers sometimes explicitly say that the the patch it to be applied to the specific tree. This information is not used in this initial PR, but will be important for the further series triage implementations.